tuning/data/setup_dataprocessor.py

```diff
@@ -613,20 +624,80 @@ def process_dataargs(
     # Note: This check should not be removed.
     #       Its important to recompute this post handling to
     #       check if we already tokenized the dataset or not.
-    is_tokenized_dataset = is_pretokenized_dataset(train_dataset or eval_dataset)
+    if odm_config:
+        is_tokenized_dataset = True
+    else:
+        is_tokenized_dataset = is_pretokenized_dataset(train_dataset or eval_dataset)
+
+    data_collator = None
+    if odm_config:
+        collators = {}
+        eval_collators = {}
+        for k, v in train_dataset.items():
+            is_tokenized_dataset = is_pretokenized_dataset(v)
+            collators[k] = get_data_collator(
+                train_args.packing,
+                data_args.response_template,
+                tokenizer,
+                is_tokenized_dataset,
+                max_seq_length,
+                data_args.instruction_template,
+                is_padding_free=is_padding_free,
+                processor=processor,
+            )
+            data_collator = collators[k]
+        for k, v in eval_dataset.items():
+            is_tokenized_dataset = is_pretokenized_dataset(v)
+            eval_collators[k] = get_data_collator(
+                train_args.packing,
+                data_args.response_template,
+                tokenizer,
+                is_tokenized_dataset,
+                max_seq_length,
+                data_args.instruction_template,
+                is_padding_free=is_padding_free,
+                processor=processor,
+            )
+    else:
+        data_collator = get_data_collator(
+            train_args.packing,
+            data_args.response_template,
+            tokenizer,
+            is_tokenized_dataset,
+            max_seq_length,
+            data_args.instruction_template,
+            is_padding_free=is_padding_free,
+            processor=processor,
+        )
+    dataset_kwargs = {}
+    if odm_config:
+        # Third Party
+        # pylint: disable=import-outside-toplevel
+        if not is_fms_accelerate_available(plugins="odm"):
```

Comment:
Shouldn't we do this check at the top as first thing?

Reply:
Can you mention where exactly?